File: src/bridge/injector.ts
/**
 * Input Injector
 * Injects user input from Telegram into Claude Code CLI
 */

import { spawnSync } from 'child_process';
import { EventEmitter } from 'events';
import logger from '../utils/logger.js';

/**
 * Injection method types
 */
type InjectionMethod = 'tmux' | 'none';

/**
 * Input Injector Configuration
 */
interface InjectorConfig {
  method: InjectionMethod;
  tmuxSession?: string;
  tmuxSocket?: string;  // Explicit socket path for tmux -S
}

/**
 * Validate a slash command against the character whitelist.
 * Only alphanumerics, underscores, hyphens, spaces, and forward slashes are allowed.
 * Exported for unit testing.
 */
export function isValidSlashCommand(command: string): boolean {
  if (!command || command.length === 0) return false;
  // Whitelist: letters, digits, underscore, hyphen, space, forward slash
  return /^[a-zA-Z0-9_\- \/]+$/.test(command);
}

/**
 * Input Injector Class
 * Handles sending user input from Telegram to Claude Code
 */
export class InputInjector extends EventEmitter {
  private method: InjectionMethod = 'none';
  private tmuxSession: string | null = null;
  private tmuxSocket: string | null = null;  // Socket path for explicit targeting

  constructor(config: Partial<InjectorConfig> = {}) {
    super();
    if (config.tmuxSession) {
      this.tmuxSession = config.tmuxSession;
    }
    if (config.method) {
      this.method = config.method;
    }
  }

  /**
   * Returns the socket args array for tmux commands.
   * Uses -S <path> when a socket path is configured, otherwise empty array.
   */
  private get socketArgs(): string[] {
    return this.tmuxSocket ? ['-S', this.tmuxSocket] : [];
  }

  /**
   * Detect available injection method
   */
  async detectMethod(): Promise<InjectionMethod> {
    // Check for tmux
    if (this.isTmuxAvailable()) {
      const session = this.detectTmuxSession();
      if (session) {
        this.tmuxSession = session;
        logger.info('Detected tmux session', { session });
        return 'tmux';
      }
    }

    logger.warn('No injection method available');
    return 'none';
  }

  /**
   * Initialize injector
   */
  async init(): Promise<boolean> {
    this.method = await this.detectMethod();

    if (this.method === 'none') {
      logger.warn('Input injection not available');
      return false;
    }

    logger.info('Input injector initialized', { method: this.method });
    return true;
  }

  /**
   * Inject input text into Claude Code
   */
  async inject(text: string): Promise<boolean> {
    switch (this.method) {
      case 'tmux':
        return this.injectViaTmux(text);

      default:
        logger.warn('No injection method configured');
        return false;
    }
  }

  /**
   * Validate that the tmux target pane exists
   * BUG-001 fix: Check before injection to fail fast with clear error
   * @returns { valid: true } if pane exists, { valid: false, reason: string } if not
   */
  validateTarget(): { valid: boolean; reason?: string } {
    if (!this.tmuxSession) {
      return { valid: false, reason: 'No tmux session configured' };
    }

    const result = spawnSync(
      'tmux',
      [...this.socketArgs, 'list-panes', '-t', this.tmuxSession],
      { stdio: 'pipe', encoding: 'utf8' }
    );

    if (result.status === 0) {
      return { valid: true };
    }

    return {
      valid: false,
      reason: `Pane "${this.tmuxSession}" not found. Claude may have moved to a different pane.`
    };
  }

  /**
   * Inject via tmux send-keys
   */
  private injectViaTmux(text: string): boolean {
    if (!this.tmuxSession) {
      logger.warn('No tmux session');
      return false;
    }

    // BUG-001 fix: Validate target exists before attempting injection
    const validation = this.validateTarget();
    if (!validation.valid) {
      logger.warn('Target validation failed', {
        session: this.tmuxSession,
        socket: this.tmuxSocket,
        reason: validation.reason
      });
      return false;
    }

    try {
      // Use spawnSync with argument arrays — no shell interpretation, no escaping needed.
      // The -l flag tells tmux to treat the string as literal key input.
      const sendResult = spawnSync(
        'tmux',
        [...this.socketArgs, 'send-keys', '-t', this.tmuxSession, '-l', text],
        { stdio: 'pipe', encoding: 'utf8' }
      );

      logger.debug('Running tmux send-keys', {
        session: this.tmuxSession,
        socket: this.tmuxSocket,
        textLength: text.length
      });

      if (sendResult.status !== 0) {
        throw new Error(sendResult.stderr || 'tmux send-keys failed');
      }

      // Send Enter key separately to submit
      const enterResult = spawnSync(
        'tmux',
        [...this.socketArgs, 'send-keys', '-t', this.tmuxSession, 'Enter'],
        { stdio: 'pipe', encoding: 'utf8' }
      );

      if (enterResult.status !== 0) {
        throw new Error(enterResult.stderr || 'tmux send-keys Enter failed');
      }

      logger.debug('Injected via tmux', { session: this.tmuxSession, socket: this.tmuxSocket });
      return true;
    } catch (error: unknown) {
      const execError = error as { stderr?: string; message?: string };
      logger.error('Failed to inject via tmux', {
        error,
        stderr: execError.stderr,
        message: execError.message,
        session: this.tmuxSession,
        socket: this.tmuxSocket,
        textLength: text.length
      });
      return false;
    }
  }

  /**
   * Check if tmux is available
   */
  private isTmuxAvailable(): boolean {
    const result = spawnSync('tmux', ['-V'], { stdio: 'ignore' });
    return result.status === 0;
  }

  /**
   * Detect current tmux session
   */
  private detectTmuxSession(): string | null {
    // Check if we're inside tmux
    if (!process.env.TMUX) {
      // Look for Claude Code sessions
      return this.findClaudeCodeSession();
    }

    // Get current session name
    const result = spawnSync('tmux', ['display-message', '-p', '#S'], {
      encoding: 'utf8'
    });

    if (result.status === 0 && result.stdout) {
      return result.stdout.trim();
    }

    return null;
  }

  /**
   * Find a tmux session running Claude Code
   */
  private findClaudeCodeSession(): string | null {
    try {
      // List all tmux sessions and panes
      const panesResult = spawnSync(
        'tmux',
        ['list-panes', '-a', '-F', '#{session_name}:#{pane_current_command}'],
        { encoding: 'utf8' }
      );

      if (panesResult.status === 0 && panesResult.stdout) {
        const lines = panesResult.stdout.trim().split('\n');

        for (const line of lines) {
          const [session, command] = line.split(':');
          // Look for node/claude processes
          if (command && (command.includes('claude') || command.includes('node'))) {
            return session;
          }
        }
      }

      // Fallback: look for any session with "claude" in the name
      const sessionsResult = spawnSync(
        'tmux',
        ['list-sessions', '-F', '#{session_name}'],
        { encoding: 'utf8' }
      );

      if (sessionsResult.status === 0 && sessionsResult.stdout) {
        const sessions = sessionsResult.stdout.trim().split('\n');
        for (const session of sessions) {
          if (session.toLowerCase().includes('claude') || session.toLowerCase().includes('code')) {
            return session;
          }
        }
      }

      return null;
    } catch {
      return null;
    }
  }

  /**
   * Escape text for tmux send-keys.
   * @deprecated No longer needed with spawnSync argument arrays — tmux receives raw text
   * without shell interpretation. Kept for any external callers.
   */
  escapeTmuxText(text: string): string {
    return text
      .replace(/\\/g, '\\\\')
      .replace(/"/g, '\\"');
  }

  /**
   * Send special key
   * BUG-004 fix: Include socket flag for correct tmux server targeting
   */
  async sendKey(key: 'Enter' | 'Escape' | 'Tab' | 'Ctrl-C' | 'Ctrl-U'): Promise<boolean> {
    if (this.method !== 'tmux' || !this.tmuxSession) {
      return false;
    }

    try {
      const keyMap: Record<string, string> = {
        'Enter': 'Enter',
        'Escape': 'Escape',
        'Tab': 'Tab',
        'Ctrl-C': 'C-c',
        'Ctrl-U': 'C-u'
      };

      // BUG-004 fix: Include socket args to target correct tmux server
      const result = spawnSync(
        'tmux',
        [...this.socketArgs, 'send-keys', '-t', this.tmuxSession, keyMap[key]],
        { stdio: 'ignore' }
      );

      return result.status === 0;
    } catch (error) {
      logger.error('Failed to send key', { key, error });
      return false;
    }
  }

  /**
   * Send slash command (like /clear)
   * Validates against the character whitelist before sending.
   * Sends command text with -l (literal) flag, then Enter key separately.
   */
  async sendSlashCommand(command: string): Promise<boolean> {
    if (this.method !== 'tmux' || !this.tmuxSession) {
      return false;
    }

    // Whitelist validation — reject commands containing shell-special characters
    if (!isValidSlashCommand(command)) {
      logger.warn('Slash command rejected: unsafe characters', { command });
      return false;
    }

    try {
      // Send command text using -l (literal) flag — no shell interpretation
      const sendResult = spawnSync(
        'tmux',
        [...this.socketArgs, 'send-keys', '-t', this.tmuxSession, '-l', command],
        { stdio: 'ignore' }
      );

      if (sendResult.status !== 0) {
        throw new Error('tmux send-keys failed for slash command');
      }

      // Send Enter separately
      const enterResult = spawnSync(
        'tmux',
        [...this.socketArgs, 'send-keys', '-t', this.tmuxSession, 'Enter'],
        { stdio: 'ignore' }
      );

      return enterResult.status === 0;
    } catch (error) {
      logger.error('Failed to send slash command', { command, error });
      return false;
    }
  }

  /**
   * Get current injection method
   */
  getMethod(): InjectionMethod {
    return this.method;
  }

  /**
   * Get tmux session name
   */
  getTmuxSession(): string | null {
    return this.tmuxSession;
  }

  /**
   * Set tmux session explicitly (with optional socket path)
   */
  setTmuxSession(session: string, socket?: string): void {
    this.tmuxSession = session;
    this.tmuxSocket = socket || null;
    if (session) {
      this.method = 'tmux';
    }
  }

  /**
   * Get tmux socket path
   */
  getTmuxSocket(): string | null {
    return this.tmuxSocket;
  }
}

/**
 * Create and initialize an input injector
 */
export async function createInjector(config?: Partial<InjectorConfig>): Promise<InputInjector | null> {
  const injector = new InputInjector(config);
  const success = await injector.init();
  return success ? injector : null;
}

export default InputInjector;

File: src/injector.rs
use crate::error::Result;
use crate::types::ALLOWED_TMUX_KEYS;
use std::process::Command;

/// Input injector for sending user input from Telegram to Claude Code CLI via tmux
///
/// Security: ALL tmux commands use Command::arg() - NO shell interpolation.
/// This prevents command injection via user-controlled inputs like session names,
/// socket paths, or message text (Security fixes #1, #2, #7).
pub struct InputInjector {
    tmux_target: Option<String>,
    tmux_socket: Option<String>,
}

impl InputInjector {
    pub fn new() -> Self {
        Self {
            tmux_target: None,
            tmux_socket: None,
        }
    }

    /// Set the tmux target and optional socket path
    /// HIGH-01: Validate socket path to prevent arbitrary socket targeting
    pub fn set_target(&mut self, target: &str, socket: Option<&str>) {
        self.tmux_target = Some(target.to_string());
        self.tmux_socket = socket.and_then(|s| {
            let path = std::path::Path::new(s);
            // Reject paths with traversal components
            if s.contains("..") {
                tracing::warn!(path = %s, "Rejecting tmux socket path with '..' traversal");
                return None;
            }
            // Must be an absolute path
            if !path.is_absolute() {
                tracing::warn!(path = %s, "Rejecting non-absolute tmux socket path");
                return None;
            }
            // Reasonable length limit
            if s.len() > 256 {
                tracing::warn!(len = s.len(), "Rejecting oversized tmux socket path");
                return None;
            }
            Some(s.to_string())
        });
    }

    /// Get current tmux target
    #[allow(dead_code)]
    pub fn target(&self) -> Option<&str> {
        self.tmux_target.as_deref()
    }

    /// Check if tmux is available
    pub fn is_tmux_available() -> bool {
        Command::new("tmux")
            .arg("-V")
            .output()
            .map(|o| o.status.success())
            .unwrap_or(false)
    }

    /// Validate that the tmux target pane exists
    pub fn validate_target(&self) -> std::result::Result<(), String> {
        let target = self
            .tmux_target
            .as_deref()
            .ok_or_else(|| "No tmux session configured".to_string())?;

        // Security fix #7: tmux target passed as .arg() never interpolated
        let mut cmd = Command::new("tmux");
        if let Some(socket) = &self.tmux_socket {
            cmd.arg("-S").arg(socket);
        }
        cmd.arg("list-panes").arg("-t").arg(target);

        match cmd.output() {
            Ok(output) if output.status.success() => Ok(()),
            _ => Err(format!(
                "Pane \"{}\" not found. Claude may have moved to a different pane.",
                target
            )),
        }
    }

    /// Inject text input into the tmux pane
    /// Security fix #1: Command::new("tmux").arg() - no shell interpolation
    pub fn inject(&self, text: &str) -> Result<bool> {
        let target = match &self.tmux_target {
            Some(t) => t,
            None => return Ok(false),
        };

        // Validate target exists
        if let Err(reason) = self.validate_target() {
            tracing::warn!(%reason, "Target validation failed");
            return Ok(false);
        }

        // Send text with -l (literal mode) - tmux handles escaping
        let mut send_cmd = Command::new("tmux");
        if let Some(socket) = &self.tmux_socket {
            send_cmd.arg("-S").arg(socket);
        }
        send_cmd
            .arg("send-keys")
            .arg("-t")
            .arg(target)
            .arg("-l") // literal mode - no special key interpretation
            .arg(text);

        let output = send_cmd.output()?;
        if !output.status.success() {
            tracing::error!(
                stderr = %String::from_utf8_lossy(&output.stderr),
                "tmux send-keys failed"
            );
            return Ok(false);
        }

        // Send Enter key separately
        let mut enter_cmd = Command::new("tmux");
        if let Some(socket) = &self.tmux_socket {
            enter_cmd.arg("-S").arg(socket);
        }
        enter_cmd
            .arg("send-keys")
            .arg("-t")
            .arg(target)
            .arg("Enter");

        let output = enter_cmd.output()?;
        if !output.status.success() {
            tracing::error!(
                stderr = %String::from_utf8_lossy(&output.stderr),
                "tmux send-keys Enter failed"
            );
            return Ok(false);
        }

        tracing::debug!(%target, text_len = text.len(), "Injected text via tmux");
        Ok(true)
    }

    /// Send a special key (from whitelist only)
    /// Security fix #7: Key is validated against whitelist, passed as .arg()
    pub fn send_key(&self, key: &str) -> Result<bool> {
        let target = match &self.tmux_target {
            Some(t) => t,
            None => return Ok(false),
        };

        // Map human-readable names to tmux key names
        let tmux_key = match key {
            "Enter" => "Enter",
            "Escape" => "Escape",
            "Tab" => "Tab",
            "Ctrl-C" => "C-c",
            "Ctrl-U" => "C-u",
            "Ctrl-D" => "C-d",
            "Ctrl-L" => "C-l",
            _ => key,
        };

        // Validate against whitelist
        if !ALLOWED_TMUX_KEYS.contains(&tmux_key) {
            tracing::warn!(key = %tmux_key, "Key not in whitelist, rejecting");
            return Ok(false);
        }

        let mut cmd = Command::new("tmux");
        if let Some(socket) = &self.tmux_socket {
            cmd.arg("-S").arg(socket);
        }
        cmd.arg("send-keys").arg("-t").arg(target).arg(tmux_key);

        let output = cmd.output()?;
        if !output.status.success() {
            tracing::error!(
                stderr = %String::from_utf8_lossy(&output.stderr),
                "tmux send-keys failed for key"
            );
            return Ok(false);
        }

        tracing::debug!(%target, %key, "Sent key via tmux");
        Ok(true)
    }

    /// Send a slash command (like /clear)
    /// Security fix #1: Command text passed as .arg(), not interpolated into shell
    pub fn send_slash_command(&self, command: &str) -> Result<bool> {
        let target = match &self.tmux_target {
            Some(t) => t,
            None => return Ok(false),
        };

        // Validate command matches safe pattern: /word (alphanumeric + hyphens only)
        let cmd_body = command.strip_prefix('/').unwrap_or(command);
        if !cmd_body
            .chars()
            .all(|c| c.is_alphanumeric() || c == '-' || c == '_' || c == ' ')
        {
            tracing::warn!(
                command,
                "Slash command rejected: contains unsafe characters"
            );
            return Ok(false);
        }

        // Send command text with -l (literal mode) to prevent tmux key interpretation
        let mut cmd = Command::new("tmux");
        if let Some(socket) = &self.tmux_socket {
            cmd.arg("-S").arg(socket);
        }
        cmd.arg("send-keys")
            .arg("-t")
            .arg(target)
            .arg("-l")
            .arg(command);

        let output = cmd.output()?;
        if !output.status.success() {
            return Ok(false);
        }

        // Send Enter
        let mut enter_cmd = Command::new("tmux");
        if let Some(socket) = &self.tmux_socket {
            enter_cmd.arg("-S").arg(socket);
        }
        enter_cmd
            .arg("send-keys")
            .arg("-t")
            .arg(target)
            .arg("Enter");

        let output = enter_cmd.output()?;
        Ok(output.status.success())
    }

    /// Check if a tmux pane is alive
    pub fn is_pane_alive(target: &str, socket: Option<&str>) -> bool {
        let mut cmd = Command::new("tmux");
        if let Some(s) = socket {
            cmd.arg("-S").arg(s);
        }
        cmd.arg("list-panes").arg("-t").arg(target);
        cmd.output().map(|o| o.status.success()).unwrap_or(false)
    }

    /// Detect current tmux session from environment
    pub fn detect_tmux_session() -> Option<TmuxInfo> {
        let tmux_env = std::env::var("TMUX").ok()?;
        let socket_path = tmux_env.split(',').next().map(|s| s.to_string());

        let session = Command::new("tmux")
            .arg("display-message")
            .arg("-p")
            .arg("#S")
            .output()
            .ok()
            .and_then(|o| {
                if o.status.success() {
                    Some(String::from_utf8_lossy(&o.stdout).trim().to_string())
                } else {
                    None
                }
            })?;

        let window = Command::new("tmux")
            .arg("display-message")
            .arg("-p")
            .arg("#I")
            .output()
            .ok()
            .and_then(|o| {
                if o.status.success() {
                    Some(String::from_utf8_lossy(&o.stdout).trim().to_string())
                } else {
                    None
                }
            })
            .unwrap_or_else(|| "0".to_string());

        let pane = Command::new("tmux")
            .arg("display-message")
            .arg("-p")
            .arg("#P")
            .output()
            .ok()
            .and_then(|o| {
                if o.status.success() {
                    Some(String::from_utf8_lossy(&o.stdout).trim().to_string())
                } else {
                    None
                }
            })
            .unwrap_or_else(|| "0".to_string());

        let target = format!("{}:{}.{}", session, window, pane);

        Some(TmuxInfo {
            session,
            pane,
            target,
            socket: socket_path,
        })
    }
}

#[derive(Debug, Clone)]
#[allow(dead_code)]
pub struct TmuxInfo {
    pub session: String,
    pub pane: String,
    pub target: String,
    pub socket: Option<String>,
}

/// Get hostname safely
pub fn get_hostname() -> String {
    Command::new("hostname")
        .output()
        .ok()
        .and_then(|o| {
            if o.status.success() {
                Some(String::from_utf8_lossy(&o.stdout).trim().to_string())
            } else {
                None
            }
        })
        .unwrap_or_default()
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_key_whitelist() {
        assert!(ALLOWED_TMUX_KEYS.contains(&"Enter"));
        assert!(ALLOWED_TMUX_KEYS.contains(&"Escape"));
        assert!(ALLOWED_TMUX_KEYS.contains(&"C-c"));
        assert!(!ALLOWED_TMUX_KEYS.contains(&"rm -rf /"));
    }

    #[test]
    fn test_injector_no_target() {
        let injector = InputInjector::new();
        assert!(injector.target().is_none());
        assert!(!injector.inject("test").unwrap());
    }
}
